### PR TITLE
ORC-1151: [C++] Fix `ColumnWriter` for non-UTC Timestamp columns

### PR DESCRIPTION
**Pull request by noirello**
### What changes were proposed in this pull request?
Fix converting non UTC timestamps for statistics.

### Why are the changes needed?
Currently, the statistics for timestamp columns are incorrect, when the writer's time zone is not UTC.

### How was this patch tested?
Ran the existing test cases.

### CODE REVIEW

File: c++/src/ColumnWriter.cc
Comment (dongjoon-hyun) [inline c++/src/ColumnWriter.cc:1840]: Thank you, @noirello . Could you make a test case for this fix?

Comment (noirello) [reply c++/src/ColumnWriter.cc:1840]: Sure, added a UTC and a non UTC test case.

Comment (noirello) [reply c++/src/ColumnWriter.cc:1840]: Please, double check the timestamps in the test cases. Time zones always can be confusing.

Comment (dongjoon-hyun) [reply c++/src/ColumnWriter.cc:1840]: It's a little counter-intuitive because we only convert `secs[i]` only. Could you add some comment here about why we don't need to convert `nanos[i]`?

File: c++/test/TestTimestampStatistics.cc
Comment (dongjoon-hyun) [inline c++/test/TestTimestampStatistics.cc:110]: Thank you for adding this.

File: c++/test/TestTimestampStatistics.cc
Comment (dongjoon-hyun) [inline c++/test/TestTimestampStatistics.cc:99]: Shall we remove an empty line?

File: c++/src/ColumnWriter.cc
Comment (dongjoon-hyun) [reply c++/src/ColumnWriter.cc:1840]: cc @wgtmac , @stiga-huang too because this is a correctness issue.

Comment (wgtmac) [reply c++/src/ColumnWriter.cc:1840]: If we have set nanosecond in the test case _**WriterTest.writeTimestampWithTimezone**_, this issue may be fixed earlier.

Comment (wgtmac) [reply c++/src/ColumnWriter.cc:1840]: > Yes, it's a little counter-intuitive because we only convert `secs[i]` only. So, `Timezone.convertToUTC` doesn't care `nanos` part properly and we should not put it here?
> 
> https://github.com/apache/orc/blob/f4c7cc14b31876e3dd4934b4bfc40c05d4b8ebca/c%2B%2B/src/Timezone.cc#L604-L606
> 
> If then, can we fix it in `Timezone.convertToUTC` instead? Is there a side-effect?

I think the current fix is good enough. As time_t uses second internally, we'd better keep the contract of Timezone.convertToUTC. We may add some comment above the fix to help understanding.

Comment (dongjoon-hyun) [reply c++/src/ColumnWriter.cc:1840]: Got it. +1 for @wgtmac 's final decision.

File: c++/test/TestTimestampStatistics.cc
Comment (noirello) [reply c++/test/TestTimestampStatistics.cc]: It's done.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side. It would be great if we can get @wgtmac or @stiga-huang 's final sign-off.

**Review by wgtmac — APPROVED** [review]
LGTM. Thanks @noirello @dongjoon-hyun

**Review by stiga-huang — APPROVED** [review]
LGTM. Thanks for fixing this, @noirello !

**Review by williamhyun — COMMENTED** [review]
+1, LGTM. 
Do we need this at ORC 1.7.5?

**Comment by noirello** [comment]
> +1, LGTM. Do we need this at ORC 1.7.5?
Yes, I think this should be included in the 1.7 branch.

**Comment by dongjoon-hyun** [comment]
+1 for backporting.

**Comment by dongjoon-hyun** [comment]
Hi All. I backported this to branch-1.7.

**Pull request merged**
